src/node/txreconciliation.cpp

```diff
@@ -363,79 +363,75 @@ class TxReconciliationTracker::Impl
                 std::holds_alternative<TxReconciliationState>(recon_state->second));
     }
 
-    bool IsFanoutTarget(const Wtxid& wtxid, NodeId peer_id, bool we_initiate, double n) const EXCLUSIVE_LOCKS_REQUIRED(m_txreconciliation_mutex)
+    std::tuple<std::vector<NodeId>, std::vector<NodeId>> GetFanoutTargets(const Wtxid& wtxid, size_t inbounds_fanout_tx_relay, size_t outbounds_fanout_tx_relay) const EXCLUSIVE_LOCKS_REQUIRED(!m_txreconciliation_mutex)
     {
-        // We use the pre-determined randomness to give a consistent result per transaction, thus making sure that no transaction
-        // gets "unlucky" if every per-peer roll fails. This means that, given a `wtxid`, the ordering will always be the same,
-        // independently of what peer this is queried for. So some need to be picked eventually (as long as `n` doesn't shrink).
-        CSipHasher deterministic_randomizer{m_deterministic_randomizer};
-        deterministic_randomizer.Write(wtxid.ToUint256());
-
-        // We decide a peer is a fanout target for a given transaction deterministically at random based on two things: the wtxid of
-        // the transaction to be relayed (which we used as a seed for the randomizer) and the likelihood to fanout this transaction
-        // (which depends on how many non-erlay tx-relay peers we have, and the direction of the connection).
-        // In order to do this, we sort the peers in random order, and then we pick the top `n` peers of the resulting collection.
-        // If our chosen peer happens to be within the picked selection, we fanout to it, otherwise we reconcile.
-        const size_t targets_size = ((deterministic_randomizer.Finalize() & 0xFFFFFFFF) + uint64_t(n * 0x100000000)) >> 32;
-
-        std::vector<std::pair<uint64_t, NodeId>> best_peers;
-        best_peers.reserve(m_states.size());
+        AssertLockNotHeld(m_txreconciliation_mutex);
+        LOCK(m_txreconciliation_mutex);
 
+        // We decide whether a particular peer is a low-fanout flood target differently based on its connection direction:
+        // - for outbounds we have a fixed number of flood destinations.
+        // - for inbounds we use a fraction of all inbound peers supporting tx relay.
+        auto outbounds_target_size = (double)OUTBOUND_FANOUT_DESTINATIONS - outbounds_fanout_tx_relay;
```

Comment:
The following test would break the code if we forget the double cast (discussion [here](https://github.com/bitcoin/bitcoin/pull/30116#discussion_r1820474026)). I suggest adding it to the tests.

```
        std::tie(in_fanout_targets, out_fanout_targets) = tracker.GetFanoutTargets(Wtxid::FromUint256(frc.rand256()), /*inbounds_fanout_tx_relay=*/0, /*outbounds_fanout_tx_relay=*/100);
        BOOST_CHECK(!tracker.ShouldFanoutTo(peer_id0, in_fanout_targets, out_fanout_targets));

```

Reply:
I've added it to the initial commit and reworked it on the last one